File: src/bj1949/Main.java
package bj1949;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String input[] = br.readLine().split(" ");
        int mrSang[]=new int[2];
        for(int i=0; i<input.length; i++){
            char temp[]=new char[3];
            temp[0]=input[i].charAt(2);
            temp[1]=input[i].charAt(1);
            temp[2]=input[i].charAt(0);
            mrSang[i]=Integer.parseInt(temp[0]+""+temp[1]+""+temp[2]);
        }
        if(mrSang[0]>mrSang[1])
            System.out.println(mrSang[0]);
        else
            System.out.println(mrSang[1]);
    }
}

class Tree{
    int index;
//    Tree
}

File: kotlin/src/main/kotlin/bj1949/Main.kt
package bj1949

import java.io.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val map = mutableMapOf<Int, List<Int>>()
    val dp = Array(n + 1) { IntArray(2) }
    val people = br.readLine().split(" ").map { it.toInt() }
    repeat(n - 1) {
        val (a, b) = br.readLine().split(" ").map { it.toInt() }
        map[a] = map.getOrDefault(a, listOf()) + b
        map[b] = map.getOrDefault(b, listOf()) + a
    }
    fun dfs(v: Int, parent: Int) {
        dp[v][0] = 0
        dp[v][1] = people[v - 1]
        for (i in map[v]!!) {
            if (i == parent) continue
            dfs(i, v)
            dp[v][0] += maxOf(dp[i][0], dp[i][1])
            dp[v][1] += dp[i][0]
        }
    }
    dfs(1, 0)
    println(maxOf(dp[1][0], dp[1][1]))
}
